### PR TITLE
[#9956] Make dataBundle return the json of actual objects created in backend

### LINKED ISSUE #9956: Make dataBundle return the json of actual objects created in backend
- **Environment**: `master` branch at commit 87f9c7180f17083161986b3311cf3785e32a8494

**Description of feature/enhancement**
Currently, when the `dataBundle` object is passed to the backend server and `PutDataBundleAction` is executed, the method only returns the HTTP response code as the json file. This makes the assumption that the values in dataBundle will be exactly same as the values as the newly generated objects in the server, which is not the case. For example, the id of the objects created always gets overwritten by the backend server when it is created, leading to the original object in dataBundle and the actual object in the backend server having different id.

**Justification**
With the current behaviour of dataBundle, it is very difficult to implement the lnp tests. There are certain endpoints that requires the id of the objects, which are impossible to do with the current implementation of dataBundle. Ensuring that the dataBundle gets updated with the new values should make the lnp testing viable for these endpoints.

### PR DESCRIPTION
**Pull request by daekoon**
Fixes #9956 and #9965 

**Outline of Solution**
`putDataBundleAction` now returns the JSON of the updated data bundle as a response. Related methods have been updated to return the DataBundle object itself, and the method `persistDataBundle` now updates all the values in the dataBundle. Test classes that makes use of dataBundles have their logic modified to support this change.

This change allows test classes that uses backdoor, to have access to updated dataBundle files.

### CODE REVIEW

File: src/main/java/teammates/logic/core/DataBundleLogic.java [hunk 1]
```diff
@@ -109,18 +110,249 @@ public void persistDataBundle(DataBundle dataBundle) throws InvalidParametersExc
...
+        updateDataBundleFeedbackSessions(dataBundle, newFeedbackSessions);
+        updateDataBundleFeedbackQuestions(dataBundle, createdQuestions);
+        updateDataBundleFeedbackResponses(dataBundle, newFeedbackResponses);
+        updateDataBundleFeedbackResponseComments(dataBundle, newFeedbackResponseComments);
+
+        return dataBundle;
+
+    }
+
+    private void updateDataBundleAccounts(DataBundle dataBundle, List<AccountAttributes> newAccounts) {
```
Comment (junming403) [inline src/main/java/teammates/logic/core/DataBundleLogic.java]: seems the logic can be simplified here?
Is "adding all accounts from 'newAccounts' that already exist in oldAccounts?"

File: src/main/java/teammates/logic/core/DataBundleLogic.java [hunk 2]
```diff
@@ -109,18 +110,249 @@ public void persistDataBundle(DataBundle dataBundle) throws InvalidParametersExc
...
+            }
+        }
+        dataBundle.accounts = accounts;
+    }
+
+    private void updateDataBundleProfiles(DataBundle dataBundle, List<StudentProfileAttributes> newProfiles) {
+        Map<String, StudentProfileAttributes> oldProfiles = dataBundle.profiles;
+        Map<String, StudentProfileAttributes> profiles = new LinkedHashMap<>();
+
+        for (Map.Entry<String, StudentProfileAttributes> entry : oldProfiles.entrySet()) {
```
Comment (junming403) [inline src/main/java/teammates/logic/core/DataBundleLogic.java]: same here？

File: src/main/java/teammates/logic/core/DataBundleLogic.java [hunk 3]
```diff
@@ -109,18 +110,249 @@ public void persistDataBundle(DataBundle dataBundle) throws InvalidParametersExc
...
+            }
+        }
+        dataBundle.profiles = profiles;
+    }
+
+    private void updateDataBundleCourses(DataBundle dataBundle, List<CourseAttributes> newCourses) {
+        Map<String, CourseAttributes> oldCourses = dataBundle.courses;
+        Map<String, CourseAttributes> courses = new LinkedHashMap<>();
+
+        for (Map.Entry<String, CourseAttributes> entry : oldCourses.entrySet()) {
```
Comment (junming403) [inline src/main/java/teammates/logic/core/DataBundleLogic.java]: same here

File: src/main/java/teammates/logic/core/DataBundleLogic.java [hunk 4]
```diff
@@ -109,18 +110,249 @@ public void persistDataBundle(DataBundle dataBundle) throws InvalidParametersExc
...
+            }
+        }
+        dataBundle.courses = courses;
+    }
+
+    private void updateDataBundleInstructors(DataBundle dataBundle, List<InstructorAttributes> newInstructors) {
+        Map<String, InstructorAttributes> oldInstructors = dataBundle.instructors;
+        Map<String, InstructorAttributes> instructors = new LinkedHashMap<>();
+
+        for (Map.Entry<String, InstructorAttributes> entry : oldInstructors.entrySet()) {
```
Comment (junming403) [inline src/main/java/teammates/logic/core/DataBundleLogic.java]: same here

File: src/main/java/teammates/logic/core/DataBundleLogic.java [hunk 5]
```diff
@@ -109,18 +110,249 @@ public void persistDataBundle(DataBundle dataBundle) throws InvalidParametersExc
...
+    private void updateDataBundleFeedbackSessions(DataBundle dataBundle,
+            List<FeedbackSessionAttributes> newFeedbackSessions) {
+        Map<String, FeedbackSessionAttributes> oldFeedbackSessions = dataBundle.feedbackSessions;
+        Map<String, FeedbackSessionAttributes> feedbackSessions = new LinkedHashMap<>();
+
+        for (Map.Entry<String, FeedbackSessionAttributes> entry : oldFeedbackSessions.entrySet()) {
+            String key = entry.getKey();
+            FeedbackSessionAttributes value = entry.getValue();
+
+            for (FeedbackSessionAttributes cur : newFeedbackSessions) {
```
Comment (junming403) [inline src/main/java/teammates/logic/core/DataBundleLogic.java]: same here

File: src/main/java/teammates/logic/core/DataBundleLogic.java [hunk 6]
```diff
@@ -109,18 +110,249 @@ public void persistDataBundle(DataBundle dataBundle) throws InvalidParametersExc
...
+        }
+        dataBundle.feedbackSessions = feedbackSessions;
+    }
+
+    private void updateDataBundleFeedbackQuestions(DataBundle dataBundle,
+                List<FeedbackQuestionAttributes> createdQuestions) {
+        Map<String, FeedbackQuestionAttributes> oldFeedbackQuestions = dataBundle.feedbackQuestions;
+        Map<String, FeedbackQuestionAttributes> feedbackQuestions = new LinkedHashMap<>();
+
+        for (Map.Entry<String, FeedbackQuestionAttributes> entry : oldFeedbackQuestions.entrySet()) {
```
Comment (junming403) [inline src/main/java/teammates/logic/core/DataBundleLogic.java]: same here

File: src/main/java/teammates/logic/core/DataBundleLogic.java [hunk 7]
```diff
@@ -109,18 +110,249 @@ public void persistDataBundle(DataBundle dataBundle) throws InvalidParametersExc
...
+    }
+
+    private void updateDataBundleFeedbackResponses(DataBundle dataBundle,
+                List<FeedbackResponseAttributes> newFeedbackResponses) {
+        Map<String, FeedbackResponseAttributes> oldFeedbackResponses = dataBundle.feedbackResponses;
+        Map<String, FeedbackResponseAttributes> feedbackResponses = new LinkedHashMap<>();
+
+        for (Map.Entry<String, FeedbackResponseAttributes> entry : oldFeedbackResponses.entrySet()) {
+            String key = entry.getKey();
+            FeedbackResponseAttributes value = entry.getValue();
```
Comment (junming403) [inline src/main/java/teammates/logic/core/DataBundleLogic.java]: same here

File: src/main/java/teammates/logic/core/DataBundleLogic.java (hunk 1 above)
Comment (daekoon) [reply src/main/java/teammates/logic/core/DataBundleLogic.java]: It is mapping back the accounts in newAccounts back to their original keys, by finding an corresponding account in the oldAccounts (which is the account before being added into the server)

The oldAccounts contain the Map<string, AccountAttributes> which contains the attributes before they were added into the server. NewAccounts contain a List of AccountAttributes, that contain the same objects as oldAccounts, but taken directly from the database after being added, which means that they have updated values that their respective counterparts in oldAccounts do not have.

Since dataBundle uses Map to store the respective attributes, when we need to update the Attributes from the new ones directly from the database, we need to ensure that they get mapped back to the correct Key. This is because the external classes that uses databundle uses the keys to retrieve the attributes, and having the wrong attributes mapped to wrong keys will result in problems for those external classes.

The first problem is that when the values are passed into the database, key-value mapping is lost and we only have the list of attributes to work on. The second issue is that the old and new value aren't exactly equal, because the new attributes have some variables that are generated by the server (createdAt, and feedbackQuestionId are such variables). The third issue is that there seems to be no requirement for uniqueness for the attribute's variables, which means that there is no single variable that can be used to link the new attributes to the old attribute.

I ended up with writing this n + 1 query to ensure that the updated attributes gets mapped back to the correct key, by comparing all possible relevant variables with the old values, that does not get changed by the database. I'm definitely open to using another approach if a better one exists.

Comment (junming403) [reply src/main/java/teammates/logic/core/DataBundleLogic.java]: In this case if you want to check if one entry in oldAccounts exists in newAccounts, probably hasing is a better idea? Currently the double nested loop seems to be not effiecit (O(n^2))

Comment (daekoon) [reply src/main/java/teammates/logic/core/DataBundleLogic.java]: Updated logic to use hashing

File: src/e2e/java/teammates/e2e/util/BackDoor.java [hunk 8]
```diff
@@ -215,8 +215,10 @@ public static String removeAndRestoreDataBundle(DataBundle dataBundle) {
         removeDataBundle(dataBundle);
         ResponseBodyAndCode putRequestOutput =
                 executePostRequest(Const.ResourceURIs.DATABUNDLE, null, JsonUtils.toJson(dataBundle));
-        return putRequestOutput.responseCode == HttpStatus.SC_OK
-                ? Const.StatusCodes.BACKDOOR_STATUS_SUCCESS : Const.StatusCodes.BACKDOOR_STATUS_FAILURE;
+        if (putRequestOutput.responseCode == HttpStatus.SC_OK) {
+            return putRequestOutput.responseBody;
```
Comment (xpdavid) [inline src/e2e/java/teammates/e2e/util/BackDoor.java]: The function return two different things (json response or `BACKDOOR_STATUS_FAILURE `) with different meanings which result in the hack in `https://github.com/TEAMMATES/teammates/pull/9967/files#diff-6509e6ed4f2a3ea02c0d46c61b1ff10fR185`. I think it is better to just throw exception for failing situation.

File: src/main/java/teammates/logic/core/DataBundleLogic.java [hunk 9]
```diff
@@ -109,18 +110,226 @@ public void persistDataBundle(DataBundle dataBundle) throws InvalidParametersExc
...
+
+        updateDataBundleProfiles(dataBundle, newProfiles);
+        updateDataBundleCourses(dataBundle, newCourses);
+        updateDataBundleInstructors(dataBundle, newInstructors);
+
+        updateDataBundleStudents(dataBundle, newStudents);
+        updateDataBundleFeedbackSessions(dataBundle, newFeedbackSessions);
+        updateDataBundleFeedbackQuestions(dataBundle, createdQuestions);
+        updateDataBundleFeedbackResponses(dataBundle, newFeedbackResponses);
+        updateDataBundleFeedbackResponseComments(dataBundle, newFeedbackResponseComments);
```
Comment (xpdavid) [inline src/main/java/teammates/logic/core/DataBundleLogic.java]: All the logic above looks the same? Why not writing a generic function?

File: src/e2e/java/teammates/e2e/util/BackDoor.java (hunk 8 above)
Comment (daekoon) [reply src/e2e/java/teammates/e2e/util/BackDoor.java]: Updated code to throw exception when HTTP request fails for `removeandupdateDataBundle`

File: src/main/java/teammates/logic/core/DataBundleLogic.java (hunk 9 above)
Comment (daekoon) [reply src/main/java/teammates/logic/core/DataBundleLogic.java]: Updated all updateXXX methods into a generic function

File: src/e2e/java/teammates/e2e/cases/lnp/BaseLNPTestCase.java [hunk 10]
```diff
@@ -162,18 +163,27 @@ protected void createTestData() {
...
      */
-    protected void persistTestData() {
+    protected void persistTestData() throws IOException, HttpRequestFailedException {
         DataBundle dataBundle = loadDataBundle(getJsonDataPath());
-        BackDoor.removeAndRestoreDataBundle(dataBundle);
+        String responseBody = "";
+        responseBody = BackDoor.removeAndRestoreDataBundle(dataBundle);
+
+        String pathToResultFile = createFileAndDirectory(TestProperties.LNP_TEST_DATA_FOLDER, getJsonDataPath());
+        String jsonValue = JsonUtils.parse(responseBody).getAsJsonObject().get("message").getAsString();
```
Comment (xpdavid) [inline src/e2e/java/teammates/e2e/cases/lnp/BaseLNPTestCase.java]: Just a nit: The whole message body can be the json. Don't use a specific field to hold them (magic "message" field) .

File: src/e2e/java/teammates/e2e/cases/lnp/BaseLNPTestCase.java [hunk 11]
```diff
@@ -162,18 +163,27 @@ protected void createTestData() {
...
+        try (BufferedWriter bw = Files.newBufferedWriter(Paths.get(pathToResultFile))) {
```
Comment (xpdavid) [inline src/e2e/java/teammates/e2e/cases/lnp/BaseLNPTestCase.java]: I think you can try `FileHelper.saveFile`

File: src/e2e/java/teammates/e2e/util/BackDoor.java [hunk 12]
```diff
@@ -212,12 +213,14 @@ private static void addAuthKeys(HttpRequestBase request) {
...
+    public static String removeAndRestoreDataBundle(DataBundle dataBundle) throws HttpRequestFailedException {
         removeDataBundle(dataBundle);
         ResponseBodyAndCode putRequestOutput =
                 executePostRequest(Const.ResourceURIs.DATABUNDLE, null, JsonUtils.toJson(dataBundle));
-        return putRequestOutput.responseCode == HttpStatus.SC_OK
-                ? Const.StatusCodes.BACKDOOR_STATUS_SUCCESS : Const.StatusCodes.BACKDOOR_STATUS_FAILURE;
+        if (putRequestOutput.responseCode == HttpStatus.SC_OK) {
+            return putRequestOutput.responseBody;
+        }
+        throw new HttpRequestFailedException("Request failed with status code: " + putRequestOutput.responseCode);
```
Comment (xpdavid) [inline src/e2e/java/teammates/e2e/util/BackDoor.java]: It OK to use `IOException` I think.

File: src/main/java/teammates/common/util/CustomLogic.java [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+        // utility class
+    }
+
+    /**
+     * Custom equals method that is able to handle null objects.
+     * @param first first object to be compared
+     * @param second second object to be compared
+     * @return true of both objects are null or equal, false otherwise.
+     */
+    public static boolean customEquals(Object first, Object second) {
```
Comment (xpdavid) [inline src/main/java/teammates/common/util/CustomLogic.java]: I think you should put the method into `EntityAttributes` since it is about comparison between `attributes`. In addition, please don't call it as `customEquals`. The name is not constructive as how "custom" is "custom". A name such as `equalsWithNullComparaision` (of course it is not a ideal name) is far better than `customEquals.`

File: src/main/java/teammates/common/datatransfer/attributes/FeedbackResponseAttributes.java [hunk 14]
```diff
@@ -168,6 +170,28 @@ public String toString() {
...
+    @Override
+    public int hashCode() {
+        StringBuilder stringBuilder = new StringBuilder();
+        stringBuilder.append(this.feedbackSessionName).append(this.courseId)
+                .append(this.feedbackQuestionId).append(this.giver).append(this.recipient);
+        return stringBuilder.toString().hashCode();
+    }
+
+    @Override
+    public boolean equals(Object other) {
```
Comment (xpdavid) [inline src/main/java/teammates/common/datatransfer/attributes/FeedbackResponseAttributes.java]: Don't forget to write test for the equals method.

File: src/e2e/java/teammates/e2e/util/BackDoor.java (hunk 12 above)
Comment (daekoon) [reply src/e2e/java/teammates/e2e/util/BackDoor.java]: I felt that Http requests should use a separate exception because they aren't exactly IO, and you might want to handle them separately if you need to.

File: src/e2e/java/teammates/e2e/cases/lnp/BaseLNPTestCase.java (hunk 10 above)
Comment (daekoon) [reply src/e2e/java/teammates/e2e/cases/lnp/BaseLNPTestCase.java]: From what I checked, in order to be able to return the entire `DataBundle` object as json I need to create a new `DataBundleData` class that extends `ApiOutput`, and converts all the values into `xxxData`. I checked through the various `xxxData` classes, and my concern is that some of the classes does not have the same structure as their respective `xxxAttributes`.

For example, the class `FeedbackQuestionData` converts the two fields, `showResponsesTo` and `showGiverNameTo`, which are `FeedbackParticipantType` class, into `FeedbackVisibilityType` class. I'm worried that using `xxxData` will have unintended side effect - should we proceed with this? (Removing the magic "message" field)

Comment (xpdavid) [reply src/e2e/java/teammates/e2e/cases/lnp/BaseLNPTestCase.java]: I am not suggestion to use `xxxData`. Let me rephrase the question, why there is `message` key here?

Comment (daekoon) [reply src/e2e/java/teammates/e2e/cases/lnp/BaseLNPTestCase.java]: This is a behaviour from the `PutDataBundleAction.java`, method execute(). 
[Link to code](https://github.com/TEAMMATES/teammates/blob/c2293eec588f0a11399ec8c02ec8cce57318f376/src/main/java/teammates/ui/webapi/action/PutDataBundleAction.java#L27-L41)

The execute() method has to return a `JsonResult` object. `JsonResult` class has 4 constructors : 
```
public JsonResult(ApiOutput output) { }
public JsonResult(ApiOutput output, List<Cookie> cookies) { }
public JsonResult(String message) { }
public JsonResult(String message, int statusCode) {}
```
As for now I'm returning the String converted, it will use the 3rd constructor which creates a `MessageOutput` object with my json String as the message. Which is why the `message` key is used.

In order to use the `ApiOutput` I need to create the `DataBundleData` class, but now that I think about it the message body will still be returned under the "DataBundle" key even if I do that.

I've tried looking around to find a way to remove the extra key (basically pass the entire json string by itself), but I couldn't find it so far. Not sure whether creating an entire group of extra classes is worth it either.

File: src/e2e/java/teammates/e2e/cases/lnp/BaseLNPTestCase.java (hunk 11 above)
Comment (daekoon) [reply src/e2e/java/teammates/e2e/cases/lnp/BaseLNPTestCase.java]: Updated method to use filehelper

File: src/main/java/teammates/common/util/CustomLogic.java (hunk 13 above)
Comment (daekoon) [reply src/main/java/teammates/common/util/CustomLogic.java]: Moved custom equals method to `EntityAttributes`, and renamed it to reflect its purpose more clearly

File: src/main/java/teammates/common/datatransfer/attributes/FeedbackResponseAttributes.java (hunk 14 above)
Comment (daekoon) [reply src/main/java/teammates/common/datatransfer/attributes/FeedbackResponseAttributes.java]: Added tests for the equals method

File: src/e2e/java/teammates/e2e/cases/lnp/BaseLNPTestCase.java [hunk 15]
```diff
@@ -162,18 +164,24 @@ protected void createTestData() {
...
      */
-    protected void persistTestData() {
+    protected void persistTestData() throws IOException, HttpRequestFailedException {
         DataBundle dataBundle = loadDataBundle(getJsonDataPath());
-        BackDoor.removeAndRestoreDataBundle(dataBundle);
+        String responseBody = "";
+        responseBody = BackDoor.removeAndRestoreDataBundle(dataBundle);
+
+        String pathToResultFile = createFileAndDirectory(TestProperties.LNP_TEST_DATA_FOLDER, getJsonDataPath());
+        String jsonValue = JsonUtils.parse(responseBody).getAsJsonObject().get("message").getAsString();
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/cases/lnp/BaseLNPTestCase.java]: I think there is a way such that this (usage of magic string `message`) can be mitigated. Let me get back to you later.

File: src/main/java/teammates/common/datatransfer/attributes/EntityAttributes.java [hunk 16]
```diff
@@ -55,6 +55,21 @@ public void addNonEmptyError(String error, List<String> errors) {
         errors.add(error);
     }
 
+    /**
+     * Custom equals method that is able to handle null objects.
+     * @param first first object to be compared
+     * @param second second object to be compared
+     * @return true of both objects are null or equal, false otherwise.
+     */
+    protected boolean nullableDataEquals(Object first, Object second) {
```
Comment (xpdavid) [inline src/main/java/teammates/common/datatransfer/attributes/EntityAttributes.java]: Actually, you can use `Objects.equals` from `java.util.Objects`

File: src/test/java/teammates/test/cases/datatransfer/StudentProfileAttributesTest.java [hunk 17]
```diff
@@ -295,6 +295,35 @@ public void testUpdateOptionsBuilder_withNullInput_shouldFailWithAssertionError(
                         .withPictureKey(null));
     }
 
+    @Test
+    public void testEquals() {
```
Comment (xpdavid) [inline src/test/java/teammates/test/cases/datatransfer/StudentProfileAttributesTest.java]: You can test: `hashCode()` of objects should be equal if there are equal.

File: src/main/java/teammates/common/datatransfer/attributes/EntityAttributes.java (hunk 16 above)
Comment (daekoon) [reply src/main/java/teammates/common/datatransfer/attributes/EntityAttributes.java]: Replaced all usage of `nullableDataEquals` with `Objects.equals`

File: src/test/java/teammates/test/cases/datatransfer/StudentProfileAttributesTest.java (hunk 17 above)
Comment (daekoon) [reply src/test/java/teammates/test/cases/datatransfer/StudentProfileAttributesTest.java]: Added tests for `hashCode()`

File: src/main/java/teammates/common/datatransfer/attributes/EntityAttributes.java (hunk 16 above)
Comment (xpdavid) [reply src/main/java/teammates/common/datatransfer/attributes/EntityAttributes.java]: You can remove the method also?
